### ISSUE TITLE
Consider all arguments used if `arguments` is used.

### ISSUE THREAD

**Issue by rulatir** [eslint/eslint#13129]
**What rule do you want to change?**

no-unused-vars

**Does this change cause the rule to produce more or fewer warnings?**

fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

new default behavior

**Please provide some example code that this change will affect:**

```js
function buildARNString(partition, service, region, accountID, resourceInfo) {
    return ARN.ify(["arn", ...arguments].join(':')).toString();
}
```

**What does the rule currently do for this code?**

Warns about unused arguments (all of them).

**What will the rule do after it's changed?**

No warnings about unused arguments, because they are actually used via `arguments`.

**Are you willing to submit a pull request to implement this change?**

Not for now, zero experience with ESLint rule development.

**Comment by anikethsaha** [comment]
I do think it is a legit requirement for strict mode. As `arguments` is available only in strict mode. 
not sure though,

+1 for this

**Comment by nzakas** [comment]
I'm not in favor of this proposal. The purpose of the rule is to let you know that a defined variable hasn't been used, and in this case, the warning is correct. If you'd like to use this pattern, I'd suggest using a disable comment to turn off the warning just for that function.

**Comment by rulatir** [comment]
> The purpose of the rule is to let you know that a defined variable hasn't been used

Function parameters are not exactly the same thing as defined variables. For example, when you override a method in a subclass, you are **forced** to define all the parameters even if you don't have any need for them (because your override may well be a no-op). Not using a function argument is significantly less likely to be a mistake than not using a variable explicitly defined in a function block. And when you merely forward the arguments with `arguments` anyway, then their values effectively **are** used, and the parameters are explicitly defined only because the language absolutely requires you to do so. Lint should never berate you for doing something that you are absolutely required to do.

**Comment by ljharb** [comment]
@anikethsaha `arguments` is most certainly available in both sloppy and strict mode.

@nzakas I do have a number of polyfill cases where I define arguments solely to set the `length` (in engines where it's not configurable), but i never reference them, because i pass all the arguments via `arguments`. An option to allow this case would relieve me of having to use an eslint override comment.

**Comment by nzakas** [comment]
The disable comments are designed for the purpose of relieving us of the need to accommodate every possible exception for rules. We have to draw a line at some point for each rule, and I think this is the line for this rule.

If others on the team feel differently I’ll defer to consensus.

**Comment by anikethsaha** [comment]
> The disable comments are designed for the purpose of relieving us of the need to accommodate every possible exception for rules. We have to draw a line at some point for each rule, and I think this is the line for this rule

@nzakas , I do agree with this. But I think to implement this, the rule doesn't need many changes.

> The purpose of the rule is to let you know that a defined variable hasn't been used, and in this case, the warning is correct. 

in a way, this is kind of a user-defined cause the function arguments are user-defined in function declaration, right!

> @anikethsaha `arguments` is most certainly available in both sloppy and strict mode.

Thanks for clearing it 👍

**Comment by rulatir** [comment]
You absolutely must define parameters when you override a method. Lint tools should never berate you for what you absolutely must do. Full stop.

**Comment by mdjermanovic** [comment]
> Every time a lint tool is about to flag something as "unused", it should ask itself: if that "unused" thing were to be removed from the program text, would the program still be valid? If the answer is negative, then the lint tool MUST NOT flag the thing as "unused", because it is in fact "used" to ensure language conformance.

Can you please provide an example where the parameters are mandatory, i.e., where removing unused params produces invalid code?

**Comment by rulatir** [comment]
> Can you please provide an example where the parameters are mandatory, i.e., where removing unused params produces invalid code?

Sure, when overriding a method in typescript. I understand that since tslint is deprecated in favor of eslint, eslint is now supposed to support typescript constructs correctly, which includes recognizing things that are mandatory.

**Comment by mdjermanovic** [comment]
> Sure, when overriding a method in typescript. I understand that since tslint is deprecated in favor of eslint, eslint is now supposed to support typescript constructs correctly, which includes recognizing things that are mandatory.

Thanks for clarifying this!

For typescript code, you should use [`@typescript-eslint/no-unused-vars`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unused-vars.md) or [`@typescript-eslint/no-unused-vars-experimental`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unused-vars-experimental.md), instead of the core `no-unused-vars` rule which isn't aware of ts specifics.

[`@typescript-eslint/eslint-plugin`](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin) extends some core rules to provide support for ts code.

**Comment by rulatir** [comment]
> For typescript code

How about pure javascript: you override a class method and the override only has any use of the _third_ argument. In that case you **must** define the first and second argument too, just to bind the argument you are interested in to the correct position in the parameter list.

**Comment by mdjermanovic** [comment]
> How about pure javascript: you override a class method and the override only has any use for the _third_ argument. In that case you **must** define the first and second argument too, just to bind the argument you are interested in to the correct position in the parameter list.

This case is already supported with the [`{ "args": "after-used" }`](https://eslint.org/docs/rules/no-unused-vars#args-after-used) option:

```js
/*eslint no-unused-vars: ["error", { "args": "after-used" }]*/

class B extends A {
  
  // ok
  foo(a, b, c) {
    return c;
  }
  
}
```

[online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXZhcnM6IFtcImVycm9yXCIsIHsgXCJhcmdzXCI6IFwiYWZ0ZXItdXNlZFwiIH1dKi9cblxuY2xhc3MgQiBleHRlbmRzIEEge1xuICBcbiAgLy8gb2tcbiAgZm9vKGEsIGIsIGMpIHtcbiAgICByZXR1cm4gYztcbiAgfVxuICBcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

There is also [`{ "args": "none" }`](https://eslint.org/docs/rules/no-unused-vars#args-none) to completely ignore params:

```js
/*eslint no-unused-vars: ["error", { "args": "none" }]*/

class B extends A {
  
  // ok
  foo(a, b, c) {}
  
}
```

[online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXZhcnM6IFtcImVycm9yXCIsIHsgXCJhcmdzXCI6IFwibm9uZVwiIH1dKi9cblxuY2xhc3MgQiBleHRlbmRzIEEge1xuICBcbiAgLy8gb2tcbiAgZm9vKGEsIGIsIGMpIHt9XG4gIFxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

**Comment by nzakas** [comment]
> @nzakas , I do agree with this. But I think to implement this, the rule doesn't need many changes.

@anikethsaha this isn’t how we decide what belongs in a rule. There are a lot of requests that would be simple to implement but would make the rule go beyond its intended scope or be an edge case that isn’t important enough to include.
